packages/contracts-bedrock/src/L1/DataAvailabilityChallenge.sol

```diff
@@ -0,0 +1,343 @@
+
+        // emit an event to notify that the challenge status is now active
+        emit ChallengeStatusChanged(challengedHash, challengedBlockNumber, ChallengeStatus.Active);
+    }
+
+    /// @notice Resolve a challenge by providing the pre-image data of the challenged commitment.
+    /// @dev The provided pre-image data is hashed (keccak256) to verify that it matches the challenged commitment.
+    ///      The function reverts if the challenge is not active or if the resolve window is not open.
+    /// @param challengedBlockNumber The block number at which the commitment was made.
+    /// @param preImage The pre-image data corresponding to the challenged commitment.
+    function resolve(uint256 challengedBlockNumber, bytes32 challengedHash, bytes calldata preImage) external {
+        // require the provided input data to match the commitment
+        if (challengedHash != keccak256(preImage)) {
+            revert InvalidInputData(keccak256(preImage), challengedHash);
+        }
+
+        // require the challenge to be active (started, not resolved, and resolve window still open)
+        if (getChallengeStatus(challengedBlockNumber, challengedHash) != ChallengeStatus.Active) {
+            revert ChallengeNotActive();
+        }
+
+        // store the block number at which the challenge was resolved
+        Challenge storage activeChallenge = challenges[challengedBlockNumber][challengedHash];
+        activeChallenge.resolvedBlock = block.number;
+
+        // emit an event to notify that the challenge status is now resolved
+        emit ChallengeStatusChanged(challengedHash, challengedBlockNumber, ChallengeStatus.Resolved);
+
+        // distribute the bond among challenger, resolver and address(0)
+        _distributeBond(activeChallenge, preImage.length, msg.sender);
+    }
+
+    /// @notice Distribute the bond of a resolved challenge among the resolver, challenger and address(0).
+    ///         The challenger is refunded the bond amount exceeding the resolution cost.
+    ///         The resolver is refunded a percentage of the resolution cost based on the `resolverRefundPercentage`
+    /// state variable.
+    ///         The remaining bond is burned by sending it to address(0).
+    /// @dev The resolution cost is approximated based on a fixed cost and variable cost depending on the size of the
+    /// pre-image.
+    ///      The real resolution cost might vary, because calldata is priced differently for zero and non-zero bytes.
+    ///      Computing the exact cost adds too much gas overhead to be worth the tradeoff.
+    /// @param resolvedChallenge The resolved challenge in storage.
+    /// @param preImageLength The size of the pre-image used to resolve the challenge.
+    /// @param resolver The address of the resolver.
+    function _distributeBond(Challenge storage resolvedChallenge, uint256 preImageLength, address resolver) internal {
```

Comment:
I see that its `storage resolvedChallenge` - there should be at least 1 failing test if we changed that to `memory resolvedChallenge`

Reply:
Could you elaborate on this? The reason it's `storage` is mostly to avoid having to load the entire challenge from storage (if we changed it to memory) and to avoid having to pass the `lockedBond` and `challenger` params separately (although we could do the latter, no strong preference here)

Reply:
i am just saying that before i approve this, i will do a manual mutation test on this function. This pattern has resulted in bugs before where `storage` and `memory` are mixed up, resulting in changes not persisting

Reply:
if it was `memory` then the following line from the code would not use sstore

```
        resolvedChallenge.lockedBond = 0;
```

Reply:
gotcha!

Reply:
`testResolveSuccess` fails if this is changed to `memory`